param/parameterized.py

```diff
@@ -424,6 +520,35 @@ def _set_instantiate(self,instantiate):
             self.instantiate = instantiate or self.constant # pylint: disable-msg=W0201
 
 
+    # TODO: quick trick to allow subscription to the setting of
+    # parameter metadata. ParameterParameter?
+
+    # Note that unlike with parameter value setting, there's no access
+    # to the Parameterized instance, so no per-instance subscription.
+
+    def __setattr__(self,name,value):        
+        old = getattr(self,name) if (name!="default" and hasattr(self,'subscribers') and name in self.subscribers) else NotImplemented
+
+        super(Parameter, self).__setattr__(name, value)
+        
+        if old is not NotImplemented:
+            for subscriber in self.subscribers[name]:
+                subscriber(Change(what=name,attribute=self._attrib_name,obj=None,cls=self._owner,old=old,new=value))
+
```

Comment:
Confused, a parameter that isn't actually a ``Parameter`` is an attribute.

Reply:
Right; this is part of setattr, which handles Parameter values and regular attributes both.

Reply:
We can't detect changes to regular attributes though, afaik `Change` can only propagate to callbacks for parameters, not attributes.

Reply:
@jlstevens is right, in that `Change` can only be about parameters of a parameterized (instance or class), not about regular attributes of a parameterized. 

Just in case the limited view on github+my coding is confusing people...this hacky `__setattr__` method is to deal with changes to attributes of a *Parameter* object (maybe we call those things parameter metadata). So, *not* about parameters and attributes of a parameterized object. The naming is really not great. `what` refers to e.g. `value`, `bounds`, `constant`, etc. `attribute` refers to the name which the owning Parameterized has for the Parameter. And `obj` and `cls` are necessary on a `Change` in general because it could come from a parameterized instance or class.

(My opinion...I would still prefer not to review the code changes yet. The `Change` object is probably not exactly what I'd like, and the `setattr` method is definitely not what I'd like, but I'd prefer to settle the functionality/behavior we want before investing time in the underpinnings. E.g. we may end up not caring about parameter metadata...or at the other extreme, we may end up wanting it to be per instance (whereas it's only per class). Having said all that, I should clean up all the issues and PRs relating to this stuff, so we have one place to discuss, because you may already have started to decide about some of these things, given you're using them...)

Reply:
Oh, I just realized why you might be looking at this stuff - I didn't see #259 until now. You're brave! :)

Reply:
Maybe keep all code comments to #253, since that what got merged, and is where most such comments are. (Some stuff from this PR is not/will not be merged.)

Reply:
@ceball Thanks for the explanation. The change to ``__setattr__`` on `Parameters`s makes sense given those are attribute changes you can look out for (i.e `what`). It sounds like you agree that `attribute` really has to be a parameter (or to be precise, a parameter name).

I share your feelings about wanting to hold back so we can refine and polish things better. That said, I do think the parallel setting (#259) will be important and therefore something to consider now.

I *think* I found a way to implement that in a way that is largely orthogonal to everything else which means I have opted to hold back on making any other code comments right now (which I agree belong in #253).